File: Java_MVP/app/src/main/java/com/gotoinc/requesin/view/home/UsersDiffUtilCallback.java
package com.gotoinc.requesin.view.home;

import com.gotoinc.requesin.view._model.User;

import java.util.List;

import androidx.recyclerview.widget.DiffUtil;

/**
 * Created by Illia Derevianko on 27.11.18.
 * GoTo Inc.
 */
public class UsersDiffUtilCallback extends DiffUtil.Callback {
    private final List<User> oldList;
    private final List<User> newList;

    public UsersDiffUtilCallback(List<User> oldList, List<User> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        User oldUser = oldList.get(oldItemPosition);
        User newUser = newList.get(newItemPosition);
        return oldUser.getId() == newUser.getId();
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        User oldUser = oldList.get(oldItemPosition);
        User newUser = newList.get(newItemPosition);
        return newUser.getFullname().equals(oldUser.getFullname())
                && newUser.getAvatar().equals(oldUser.getAvatar());
    }
}

File: Kotlin_MVP/app/src/main/java/com/gotoinc/requesin/view/home/UsersDiffUtilCallback.kt
package com.gotoinc.requesin.view.home

import androidx.recyclerview.widget.DiffUtil
import com.gotoinc.requesin.view.model.User

/**
 * Created by Illia Derevianko on 29.11.18.
 * GoTo Inc.
 */
class UsersDiffUtilCallback(private val oldList: List<User>, private val newList: List<User>) : DiffUtil.Callback() {
    override fun getOldListSize() = oldList.size
    override fun getNewListSize() = newList.size

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldUser: User = oldList[oldItemPosition]
        val newUser: User = newList[newItemPosition]

        return oldUser.id == newUser.id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldUser: User = oldList[oldItemPosition]
        val newUser: User = newList[newItemPosition]

        return newUser.fullName.equals(oldUser.fullName)
                && newUser.avatar.equals(oldUser.avatar)
    }
}